cmd/artifactPrepareVersion.go

```diff
@@ -0,0 +1,351 @@
+		}
+	}
+
+	versioningType := config.VersioningType
+
+	// support former groovy versioning template and translate into new options
+	if len(config.VersioningTemplate) > 0 {
+		versioningType, _, config.IncludeCommitID = templateCompatibility(config.VersioningTemplate)
+	}
+
+	version, err := artifact.GetVersion()
+	if err != nil {
+		return errors.Wrap(err, "failed to retrieve version")
+	}
+	log.Entry().Infof("Version before automatic versioning: %v", version)
+
+	gitCommit, err := getGitCommitID(repository)
+	if err != nil {
+		return err
+	}
+	gitCommitID := gitCommit.String()
+
+	newVersion := version
+
+	if versioningType == "cloud" {
+		versioningTempl, err := versioningTemplate(artifact.VersioningScheme())
+		if err != nil {
+			return errors.Wrapf(err, "failed to get versioning template for scheme '%v'", artifact.VersioningScheme())
+		}
+
+		now := time.Now()
+
+		newVersion, err = calculateNewVersion(versioningTempl, version, gitCommitID, config.IncludeCommitID, now)
+		if err != nil {
+			return errors.Wrap(err, "failed to calculate new version")
+		}
+
+		worktree, err := getWorktree(repository)
+		if err != nil {
+			return errors.Wrap(err, "failed to retrieve git worktree")
+		}
+
+		// opening repository does not seem to consider already existing files properly
+		// behavior in case we do not run initializeWorktree:
+		//   git.Add(".") will add the complete workspace instead of only changed files
+		err = initializeWorktree(gitCommit, worktree)
+		if err != nil {
+			return err
+		}
+
+		// only update version in build descriptor if required in order to save prossing time (e.g. maven case)
+		if newVersion != version {
+			err = artifact.SetVersion(newVersion)
+			if err != nil {
+				return errors.Wrap(err, "failed to write version")
+			}
+		}
+
+		//ToDo: what about closure in current Groovy step. Discard the possibility or provide extension mechanism?
+
+		// commit changes and push to repository (including new version tag)
+		gitCommitID, err = pushChanges(config, newVersion, repository, worktree, now)
+		if err != nil {
+			return errors.Wrapf(err, "failed to push changes for version '%v'", newVersion)
+		}
+	}
+
+	log.Entry().Infof("New version: '%v'", newVersion)
+
+	commonPipelineEnvironment.git.commitID = gitCommitID
+	commonPipelineEnvironment.artifactVersion = newVersion
+
+	return nil
+}
+
+func openGit() (gitRepository, error) {
+	workdir, _ := os.Getwd()
+	return git.PlainOpen(workdir)
+}
+
+func getGitCommitID(repository gitRepository) (plumbing.Hash, error) {
+	commitID, err := repository.ResolveRevision(plumbing.Revision("HEAD"))
+	if err != nil {
+		return plumbing.Hash{}, errors.Wrap(err, "failed to retrieve git commit ID")
+	}
+	return *commitID, nil
+}
+
+func versioningTemplate(scheme string) (string, error) {
+	// generally: timestamp acts as build number providing a proper order
+	switch scheme {
+	case "maven":
+		// according to https://www.mojohaus.org/versions-maven-plugin/version-rules.html
```

Comment:
are you aware of this disclaimer?

> This has been marked deprecated and will be removed with release 3.0.0 of the versions-maven-plugin. If you are a user of this please vote on the issue #157.

might not be a good idea to lock us out of updating to future versions

Reply:
Ok, the issue mentions 

> Based on the good feedback this will not been removed.


so it should be fine